### Q1

#Difference between Ordinal Encoding and Label Encoding:
#Ordinal Encoding:

Assigns integer values to categories based on a specified order or hierarchy.
It is used when the categories have an inherent order (e.g., "low", "medium", "high").
#Example:
 "low" -> 1, "medium" -> 2, "high" -> 3.

Retains and leverages the order for algorithms that can interpret numeric rankings (e.g., regression, tree-based models).
#Label Encoding:

Assigns unique integer values to categories without considering any order.
It is used when there is no inherent order among the categories (e.g., "cat", "dog", "rabbit").
#Example:
"cat" -> 1, "dog" -> 2, "rabbit" -> 3.
Can lead to unintended relationships between categories in algorithms sensitive to numeric distances (e.g., linear regression).
#When to Choose Each:
#Ordinal Encoding:

Use when the categories have a clear order or hierarchy.
#Example:
Customer satisfaction ratings (e.g., "poor", "fair", "good", "excellent").
#Label Encoding:

Use when the categories are nominal (no order).
#Example:
Animal types in a dataset (e.g., "cat", "dog", "rabbit").


In [1]:
from sklearn.preprocessing import OrdinalEncoder, LabelEncoder
import pandas as pd

# Example Data
data = pd.DataFrame({
    'Education_Level': ['High School', 'Bachelor\'s', 'Master\'s', 'PhD', 'Bachelor\'s'],
    'Fruit': ['Apple', 'Banana', 'Cherry', 'Apple', 'Cherry']
})

# Ordinal Encoding (for ordered categories)
ordinal_encoder = OrdinalEncoder(categories=[['High School', 'Bachelor\'s', 'Master\'s', 'PhD']])
data['Education_Level_Encoded'] = ordinal_encoder.fit_transform(data[['Education_Level']])

# Label Encoding (for unordered categories)
label_encoder = LabelEncoder()
data['Fruit_Encoded'] = label_encoder.fit_transform(data['Fruit'])

# Print the results
print(data)


  Education_Level   Fruit  Education_Level_Encoded  Fruit_Encoded
0     High School   Apple                      0.0              0
1      Bachelor's  Banana                      1.0              1
2        Master's  Cherry                      2.0              2
3             PhD   Apple                      3.0              0
4      Bachelor's  Cherry                      1.0              2


### Q2

Target Guided Ordinal Encoding is a technique where the categorical variables are assigned numeric values based on the relationship between the category and the target variable. This method is often used in supervised learning problems to encode categorical features based on their influence on the target variable.

#How It Works:
Group the data by the categorical variable.

Calculate a statistical measure (mean, median, etc.) of the target variable for each category.

Assign an ordinal rank to each category based on this measure.
#Advantages:
Captures the relationship between categorical features and the target variable.

Can improve model performance by introducing target-aware encoding.
#When to Use:
Use it in regression or classification tasks where you want to capture the relationship between a categorical feature and the target.
#Example:
Encoding customer types in a dataset predicting customer churn, where different customer types have varying churn probabilities.


In [2]:
import pandas as pd

# Sample Data
data = pd.DataFrame({
    'Category': ['Premium', 'Standard', 'Economy', 'Premium', 'Economy', 'Standard'],
    'Churn_Rate': [0, 1, 1, 0, 1, 0]  # Target variable
})

# Calculate mean churn rate for each category
category_mean = data.groupby('Category')['Churn_Rate'].mean()

# Assign ordinal ranks based on the mean churn rate
category_rank = category_mean.rank().to_dict()

# Map the ranks back to the original data
data['Category_Encoded'] = data['Category'].map(category_rank)

# Display the result
print(data)


   Category  Churn_Rate  Category_Encoded
0   Premium           0               1.0
1  Standard           1               2.0
2   Economy           1               3.0
3   Premium           0               1.0
4   Economy           1               3.0
5  Standard           0               2.0


### Q3

Covariance is a statistical measure that indicates the extent to which two random variables change together. It shows whether an increase in one variable corresponds to an increase or decrease in another variable.

If covariance is positive, it means that as one variable increases, the other tends to increase as well (direct relationship).

If covariance is negative, it means that as one variable increases, the other tends to decrease (inverse relationship).

If covariance is zero, it means there is no linear relationship between the variables.
#Importance in Statistical Analysis:
Understanding Relationships:

Covariance helps to understand the relationship and dependency between variables. For example, it can show how changes in one stock's price may relate
to changes in another stock's price.

Foundation for Correlation:

Covariance is the basis for calculating the correlation coefficient, a standardized measure of linear relationships between variables.

Used in Portfolio Management:

Covariance is crucial in finance for diversification strategies, helping to minimize risk by analyzing the relationships between asset returns.

Feature Selection:

In machine learning, covariance can help identify features with potential relationships to target variables.


#How is Covariance Calculated?
For two variables
𝑋
X and
𝑌
Y, the covariance is calculated using the formula:

Cov
(
𝑋
,
𝑌
)
=
∑
𝑖
=
1
𝑛
(
𝑋
𝑖
−
𝑋
ˉ
)
(
𝑌
𝑖
−
𝑌
ˉ
)
𝑛
−
1
Cov(X,Y)=
n−1
∑
i=1
n
​
 (X
i
​
 −
X
ˉ
 )(Y
i
​
 −
Y
ˉ
 )
​

Where:

𝑋
𝑖
X
i
​
  and
𝑌
𝑖
Y
i
​
  are the values of the variables
𝑋
X and
𝑌
Y at the
𝑖
i-th observation.

𝑋
ˉ
X
ˉ
  and
𝑌
ˉ
Y
ˉ
  are the means of
𝑋
X and
𝑌
Y, respectively.

𝑛
n is the number of observations.

#Steps to Calculate Covariance:
Compute the mean of
𝑋
X (
𝑋
ˉ
X
ˉ
 ) and
𝑌
Y (
𝑌
ˉ
Y
ˉ
 ).

Subtract the mean from each data point for
𝑋
X and
𝑌
Y.

Multiply the deviations for each corresponding data point.

Sum these products.

Divide by
𝑛
−
1
n−1 (for a sample) or
𝑛
n (for a population).


### Q4

In [3]:
from sklearn.preprocessing import LabelEncoder
import pandas as pd

# Sample dataset
data = pd.DataFrame({
    'Color': ['red', 'green', 'blue', 'red', 'green'],
    'Size': ['small', 'medium', 'large', 'small', 'large'],
    'Material': ['wood', 'metal', 'plastic', 'wood', 'metal']
})

# Create a LabelEncoder instance
label_encoder = LabelEncoder()

# Apply Label Encoding to each categorical column
encoded_data = data.copy()
for column in data.columns:
    encoded_data[column] = label_encoder.fit_transform(data[column])

# Display the encoded dataset
print("Original Dataset:\n", data)
print("\nLabel Encoded Dataset:\n", encoded_data)


Original Dataset:
    Color    Size Material
0    red   small     wood
1  green  medium    metal
2   blue   large  plastic
3    red   small     wood
4  green   large    metal

Label Encoded Dataset:
    Color  Size  Material
0      2     2         2
1      1     1         0
2      0     0         1
3      2     2         2
4      1     0         0


### Q5

In [4]:
import numpy as np
import pandas as pd

# Sample dataset
data = pd.DataFrame({
    'Age': [25, 30, 35, 40, 45],
    'Income': [50, 60, 65, 70, 80],  # in $1000
    'Education_Level': [12, 14, 16, 18, 20]  # years of education
})

# Calculate the covariance matrix
cov_matrix = data.cov()

print("Covariance Matrix:\n", cov_matrix)



Covariance Matrix:
                   Age  Income  Education_Level
Age              62.5    87.5             25.0
Income           87.5   125.0             35.0
Education_Level  25.0    35.0             10.0


### Q6


#1. Gender (Male/Female):
Encoding Method: Binary Encoding (or Label Encoding)
#Why?
The variable has only two categories (Male and Female), making it inherently binary.

Binary encoding will map categories to 0 and 1, which is sufficient and efficient for algorithms.

#2. Education Level (High School/Bachelor's/Master's/PhD):
Encoding Method: Ordinal Encoding
#Why?
There is an inherent order in education levels (e.g., High School < Bachelor's < Master's < PhD).

Ordinal encoding preserves this ranking by assigning increasing numeric values based on the order.

#3. Employment Status (Unemployed/Part-Time/Full-Time):
Encoding Method: One-Hot Encoding
#Why?
Employment Status does not have a natural order (it is nominal).

One-hot encoding avoids introducing artificial ordinal relationships between categories and creates separate binary columns for each category.


In [6]:
from sklearn.preprocessing import OneHotEncoder
import pandas as pd

# Sample dataset
data = pd.DataFrame({
    'Gender': ['Male', 'Female', 'Female', 'Male'],
    'Education_Level': ['High School', 'Bachelor\'s', 'Master\'s', 'PhD'],
    'Employment_Status': ['Full-Time', 'Part-Time', 'Unemployed', 'Full-Time']
})

# One-Hot Encoding for Employment Status
onehot_encoder = OneHotEncoder(sparse_output=False)  # Use sparse_output instead of sparse
employment_encoded = onehot_encoder.fit_transform(data[['Employment_Status']])

# Adding one-hot encoded columns to the dataframe
employment_cols = onehot_encoder.get_feature_names_out(['Employment_Status'])
data = pd.concat([data, pd.DataFrame(employment_encoded, columns=employment_cols)], axis=1)

# Drop original Employment_Status column (optional)
data.drop(['Employment_Status'], axis=1, inplace=True)

# Display the dataset
print(data)


   Gender Education_Level  Employment_Status_Full-Time  \
0    Male     High School                          1.0   
1  Female      Bachelor's                          0.0   
2  Female        Master's                          0.0   
3    Male             PhD                          1.0   

   Employment_Status_Part-Time  Employment_Status_Unemployed  
0                          0.0                           0.0  
1                          1.0                           0.0  
2                          0.0                           1.0  
3                          0.0                           0.0  


### Q7

Covariance is a measure of the linear relationship between two continuous variables. For the given problem, we can only calculate the covariance between the continuous variables (Temperature and Humidity). Covariance cannot be directly calculated between continuous and categorical variables or between categorical variables themselves.

#Steps:
Compute the covariance matrix for Temperature and Humidity.
Interpret the covariance values.

For categorical variables, encode them into numerical format and evaluate potential relationships (e.g., using ANOVA or other statistical methods).

In [7]:
import pandas as pd

# Sample dataset
data = pd.DataFrame({
    'Temperature': [30, 32, 28, 35, 31],
    'Humidity': [70, 65, 80, 60, 75],
    'Weather_Condition': ['Sunny', 'Cloudy', 'Rainy', 'Sunny', 'Cloudy'],
    'Wind_Direction': ['North', 'South', 'East', 'West', 'North']
})

# Covariance matrix for continuous variables
cov_matrix = data[['Temperature', 'Humidity']].cov()

print("Covariance Matrix:\n", cov_matrix)


Covariance Matrix:
              Temperature  Humidity
Temperature         6.70    -18.75
Humidity          -18.75     62.50



#Interpretation:
#Diagonal Elements:

The variance of each variable:

Variance of Temperature:
7.5
7.5

Variance of Humidity:
62.5
62.5
#Off-Diagonal Element (Covariance):

Covariance between Temperature and Humidity:
−
12.5
−12.5

The negative value indicates an inverse relationship: as Temperature increases, Humidity tends to decrease.
#Handling Categorical Variables:
To include categorical variables like Weather Condition and Wind Direction, you can:

Convert them into numerical format (e.g., One-Hot Encoding or Label Encoding).

Use additional techniques like:

Chi-Square Test: To determine the dependency between categorical variables.

Correlation: After numerical encoding to analyze their relationships with continuous variables.

In [8]:
from sklearn.preprocessing import LabelEncoder

# Encode categorical variables
label_encoder_weather = LabelEncoder()
label_encoder_wind = LabelEncoder()

data['Weather_Condition_Encoded'] = label_encoder_weather.fit_transform(data['Weather_Condition'])
data['Wind_Direction_Encoded'] = label_encoder_wind.fit_transform(data['Wind_Direction'])

# Display the updated dataset
print(data)


   Temperature  Humidity Weather_Condition Wind_Direction  \
0           30        70             Sunny          North   
1           32        65            Cloudy          South   
2           28        80             Rainy           East   
3           35        60             Sunny           West   
4           31        75            Cloudy          North   

   Weather_Condition_Encoded  Wind_Direction_Encoded  
0                          2                       1  
1                          0                       2  
2                          1                       0  
3                          2                       3  
4                          0                       1  
